Tasks/UnityBuild/UnityBuildV3/unity-versioning.ts

```diff
@@ -0,0 +1,306 @@
+import tl = require("azure-pipelines-task-lib/task");
+import {
+  buildNumberOutputVariableName,
+  buildTargetInputVariableName,
+  bundleVersionOutputVariableName,
+  gitTagOutputVariableName,
+  projectVersioningBuildNumberModeVariableName,
+  projectVersioningBuildNumberVariableName,
+  projectVersioningBundleVersionMajorVariableName,
+  projectVersioningBundleVersionMinorVariableName,
+  projectVersioningBundleVersionModeVariableName,
+  projectVersioningBundleVersionPatchVariableName,
+  projectVersioningCommitChangesUserMailVariableName,
+  projectVersioningCommitChangesUserNameVariableName,
+  projectVersioningCommitChangesVariableName,
+  projectVersioningCreateTagPatternVariableName,
+  projectVersioningCreateTagVariableName,
+  unityProjectPathInputVariableName,
+} from "./variables";
+import {
+  SemanticVersion,
+  UnityVersioningTools,
+} from "@dinomite-studios/unity-azure-pipelines-tasks-lib";
+import { BuildPlatform } from "./build-platform";
+
+enum VersioningMode {
+  None = "none",
+  Increment = "increment",
+  Set = "set",
+}
+
+export class UnityVersioning {
+  public static run(): number {
```

Comment:
I see very clearly the need of tests because of the cyclomatic complexity here with the different values of `VersioningMode` and `BuildPlatform`.

It is very easy to create a bug or a regression when making a change in this logic.